log4j-api-java9/src/main/java/org/apache/logging/log4j/util/StackLocator.java

```diff
@@ -72,15 +75,38 @@ private StackLocator() {
     }
 
     public StackTraceElement calcLocation(final String fqcnOfLogger) {
-        return stackWalker.walk(
-                s -> s.dropWhile(f -> !f.getClassName().equals(fqcnOfLogger)) // drop the top frames until we reach the logger
-                        .dropWhile(f -> f.getClassName().equals(fqcnOfLogger)) // drop the logger frames
-                        .findFirst())
-                .get()
-                .toStackTraceElement();
+        return walker.walk(LOCATOR.get().setFqcn(fqcnOfLogger)).toStackTraceElement();
     }
 
     public StackTraceElement getStackTraceElement(final int depth) {
         return stackWalker.walk(s -> s.skip(depth).findFirst()).get().toStackTraceElement();
     }
+
+    static final class FqcnCallerLocator implements Function<Stream<StackWalker.StackFrame>, StackWalker.StackFrame> {
+
+        private String fqcn;
+
+        public FqcnCallerLocator setFqcn(String fqcn) {
+            this.fqcn = fqcn;
+            return this;
+        }
+
+        @Override
+        public StackWalker.StackFrame apply(Stream<StackWalker.StackFrame> stackFrameStream) {
+            boolean foundFqcn = false;
+            Object[] frames = stackFrameStream.toArray();
```

Comment:
Is this going to be more expensive for logging that occurs with a large stack?

Reply:
I don't know but I suspect not. The test I am running has a stack size of 35 entries. My understanding is the StackFrames are optimized for when you access less than 8., but it might just be grabbing them in batches of 8. I haven't checked that. I do know that the stream logic is adding considerable overhead. I suspect this would be even faster if there were methods that could do this inside of StackWalker without having to create the Stream.

Reply:
I see, I didn't realize it was chunked internally, that's great! It's unfortunate that streams don't JIT very well compared to standard loops/iterators, the array sounds like the way to go if it performs better than the fluent stream approach.

Reply:
Ultimately I will have to create a separate project at GitHub with JMH benchmarks to show how the variations behave and then open tickets against the JDK to see what can be done if it can be shown that using Streams isn't the most performant way to access stack traces.

Reply:
It seems we already have JMH benchmarks. I ran them. The results are in the Jira issue.